drivers: counter: mcux_rtc: allow setting a new alarm from the callback

henrikbrixandersen opened the pull request:
Allow setting a new alarm from the callback context of the just expired alarm.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
This, at least, is needed to work with the current implementation of the `counter_mcux_rtc.c`driver, which has the following check in `set_alarm()`:

```
if (data->alarm_callback != NULL) {
        return -EBUSY;
}
```

nordic-krch commented:
it's rather a bug in the driver. Once expiration callback is called, channel is considered available:

https://github.com/zephyrproject-rtos/zephyr/blob/fb0faaa1239ad5f6577d3268c67ba168aab57656/include/counter.h#L270

nordic-krch commented:
it should be cleared here:
https://github.com/zephyrproject-rtos/zephyr/blob/fb0faaa1239ad5f6577d3268c67ba168aab57656/drivers/counter/counter_mcux_rtc.c#L190
before calling user callback.

henrikbrixandersen committed:
drivers: counter: mcux_rtc: allow setting a new alarm from the callback

Allow setting a new alarm from the callback context of the just
expired alarm.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> it should be cleared here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/fb0faaa1239ad5f6577d3268c67ba168aab57656/drivers/counter/counter_mcux_rtc.c#L190
> 
> 
> before calling user callback.

@nordic-krch Thank you for that explanation. I have updated the branch to do just that.

nordic-krch reviewed (approved):
looks good

henrikbrixandersen commented:
recheck

henrikbrixandersen commented:
recheck
